olp-cpp-sdk-core/src/client/OlpClient.cpp

```diff
@@ -142,19 +148,40 @@ HttpResponse SendRequest(const http::NetworkRequest& request,
         outcome = settings.network_request_handler->Send(
             request, response_body,
             [&, interest_flag](http::NetworkResponse response) {
-              if (interest_flag->exchange(false)) {
-                network_response = std::move(response);
-                condition.Notify();
+              auto state = interest_flag->load();
+
+              // Operation has been cancelled before this function call.
+              // Just exit without processing.
+              if (state == State::Done) {
+                return;
+              }
+
+              // Headers callback may be not called at all so we
+              // still can be in waiting state. In this case try
+              // to switch to Processing state to block caller thread
+              // from exiting SendRequest() function.
+              if (state == State::Waiting) {
+                if (!interest_flag->compare_exchange_weak(state,
+                                                          State::Processing)) {
+                  return;
+                }
               }
+
+              network_response = std::move(response);
+              interest_flag->store(State::Done);
+              condition.Notify();
             },
             [&headers, interest_flag](std::string key, std::string value) {
-              if (interest_flag->load()) {
+              State state = State::Waiting;
+              if (interest_flag->compare_exchange_weak(state,
```

Comment:
This is not correct. This header callback is called multiple times, for each header once. The second time it is called the value of the interest_flag is already State::Processing so the compare_exchange_weak is going to return false so you will never reach to add any other header to the headers list.